File: src/treemapper/diffctx/edges/semantic/python.py
# pylint: disable=duplicate-code
from __future__ import annotations

import ast
import logging
import os
import re
import shutil
import subprocess
from collections import defaultdict
from pathlib import Path

from ...config.weights import LANG_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, _strip_source_prefix, path_to_module
from .python_semantics import PyFragmentInfo, analyze_python_fragment

logger = logging.getLogger(__name__)

_RG_BIN = None if os.environ.get("DIFFCTX_NO_RIPGREP") else shutil.which("rg")

_IMPORT_FROM_RE = re.compile(r"^\s*from\s+(\.+)?([\w.]*)\s+import")
_IMPORT_SIMPLE_RE = re.compile(r"^\s*import\s+([\w.]+(?:\s*,\s*[\w.]+)*)")

_INIT_PY = "__init__.py"
_PYTHON_EXTS = {".py", ".pyi", ".pyw"}

_PY_WEIGHTS = LANG_WEIGHTS["python"]
_CALL_WEIGHT = _PY_WEIGHTS.call
_SYMBOL_REF_WEIGHT = _PY_WEIGHTS.symbol_ref
_TYPE_REF_WEIGHT = _PY_WEIGHTS.type_ref


def _is_python_file(path: Path) -> bool:
    return path.suffix.lower() in _PYTHON_EXTS


def _add_import_with_prefixes(imports: set[str], imported: str) -> None:
    imports.add(imported)
    parts = imported.split(".")
    for i in range(1, len(parts) + 1):
        imports.add(".".join(parts[:i]))


def _resolve_relative(name: str, source_path: Path, repo_root: Path | None) -> str | None:
    try:
        import importlib.util

        pkg_parts = _strip_source_prefix(list(source_path.parent.parts))
        if pkg_parts and pkg_parts[-1] == "__pycache__":
            pkg_parts = pkg_parts[:-1]
        if repo_root and source_path.is_absolute():
            try:
                source_path = source_path.relative_to(repo_root)
                pkg_parts = _strip_source_prefix(list(source_path.parent.parts))
            except ValueError:
                pass
        package = ".".join(pkg_parts) if pkg_parts else None
        if not package:
            return None
        return importlib.util.resolve_name(name, package)
    except (ImportError, ValueError):
        return None


def _collect_import_from(
    node: ast.ImportFrom,
    imports: set[str],
    source_path: Path | None,
    repo_root: Path | None,
) -> None:
    module = node.module or ""
    if node.level and node.level > 0:
        relative = "." * node.level + module
        if source_path:
            resolved = _resolve_relative(relative, source_path, repo_root)
            if resolved:
                _add_import_with_prefixes(imports, resolved)
    elif module:
        _add_import_with_prefixes(imports, module)
        for alias in node.names:
            if alias.name and alias.name != "*":
                _add_import_with_prefixes(imports, f"{module}.{alias.name}")


def _extract_imports_from_content(content: str, source_path: Path | None = None, repo_root: Path | None = None) -> set[str]:
    imports: set[str] = set()
    import warnings

    try:
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", SyntaxWarning)
            tree = ast.parse(content)
    except SyntaxError:
        return imports

    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                if alias.name:
                    _add_import_with_prefixes(imports, alias.name)
        elif isinstance(node, ast.ImportFrom):
            _collect_import_from(node, imports, source_path, repo_root)
    return imports


def _resolve_relative_rg(source_path: Path, repo_root: Path, level: int, module: str) -> str | None:
    try:
        rel = source_path.relative_to(repo_root)
    except ValueError:
        return None
    parts = _strip_source_prefix(list(rel.parent.parts))
    if parts and parts[-1] == "__pycache__":
        parts = parts[:-1]
    if level > len(parts):
        return None
    base_parts = parts[: len(parts) - level + 1]
    if module:
        base_parts.extend(module.split("."))
    return ".".join(base_parts) if base_parts else None


def _build_import_index_rg(
    repo_root: Path,
    candidate_set: set[Path],
) -> dict[Path, set[str]]:
    r = subprocess.run(
        [
            _RG_BIN or "rg",
            "--no-heading",
            "--with-filename",
            r"^\s*(?:from\s+\.*[\w.]*\s+import|import\s+[\w.])",
            "--type",
            "py",
            str(repo_root),
        ],
        capture_output=True,
        text=True,
        timeout=30,
    )
    file_to_imports: dict[Path, set[str]] = defaultdict(set)
    for line in r.stdout.splitlines():
        idx = line.find(":")
        if idx < 0:
            continue
        path_str = line[:idx]
        rest = line[idx + 1 :]
        idx2 = rest.find(":")
        if idx2 >= 0 and rest[:idx2].isdigit():
            content = rest[idx2 + 1 :]
        else:
            content = rest

        path = Path(path_str)
        if path not in candidate_set:
            continue

        stripped = content.strip()
        m_from = _IMPORT_FROM_RE.match(stripped)
        m_simple = _IMPORT_SIMPLE_RE.match(stripped)
        if not m_from and not m_simple:
            continue

        if m_simple:
            for name in m_simple.group(1).split(","):
                name = name.split(" as ")[0].strip()
                if name:
                    _add_import_with_prefixes(file_to_imports[path], name)
        elif m_from:
            dots, module = m_from.group(1), m_from.group(2)
            if dots:
                resolved = _resolve_relative_rg(path, repo_root, len(dots), module or "")
                if resolved:
                    _add_import_with_prefixes(file_to_imports[path], resolved)
            elif module:
                _add_import_with_prefixes(file_to_imports[path], module)

    return file_to_imports


_REEXPORT_MAX_DEPTH = 3


def _resolve_import_target_dir(node: ast.ImportFrom, pkg_dir: Path, repo_root: Path | None) -> Path | None:
    source_module = node.module or ""
    if node.level and node.level > 0:
        target_dir = pkg_dir
        for _ in range(node.level - 1):
            target_dir = target_dir.parent
        if source_module:
            target_dir = target_dir / Path(*source_module.split("."))
        return target_dir
    if source_module and repo_root:
        return repo_root / Path(*source_module.split("."))
    return None


def _find_module_file(target_dir: Path) -> Path | None:
    source_as_file = target_dir.with_suffix(".py")
    if source_as_file.is_file():
        return source_as_file
    source_as_pkg = target_dir / _INIT_PY
    if source_as_pkg.is_file():
        return source_as_pkg
    for ext in (".pyi", ".pyw"):
        candidate = target_dir.with_suffix(ext)
        if candidate.is_file():
            return candidate
    return None


def _resolve_init_reexports(
    init_path: Path,
    repo_root: Path | None,
    file_cache: dict[Path, str] | None = None,
    _depth: int = 0,
) -> dict[str, Path]:
    if _depth >= _REEXPORT_MAX_DEPTH:
        return {}

    content = file_cache.get(init_path) if file_cache else None
    if content is None:
        try:
            content = init_path.read_text(errors="replace")
        except OSError:
            return {}

    import warnings

    try:
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", SyntaxWarning)
            tree = ast.parse(content)
    except SyntaxError:
        return {}

    result: dict[str, Path] = {}
    pkg_dir = init_path.parent

    for node in ast.walk(tree):
        if not isinstance(node, ast.ImportFrom) or not node.names:
            continue

        target_dir = _resolve_import_target_dir(node, pkg_dir, repo_root)
        if target_dir is None:
            continue

        resolved_path = _find_module_file(target_dir)
        if resolved_path is None:
            continue

        for alias in node.names:
            _name = alias.asname or alias.name
            if _name == "*":
                result[f"*:{resolved_path}"] = resolved_path
                continue
            result[_name] = resolved_path

            if resolved_path.name == _INIT_PY:
                nested = _resolve_init_reexports(resolved_path, repo_root, file_cache, _depth + 1)
                if _name in nested:
                    result[_name] = nested[_name]

    return result


class PythonEdgeBuilder(EdgeBuilder):
    weight = 0.70
    reverse_weight_factor = 0.5

    _DISCOVERY_MAX_DEPTH = 2

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
        **kwargs: object,
    ) -> list[Path]:
        py_changed = [f for f in changed_files if _is_python_file(f)]
        if not py_changed:
            return []

        fc = kwargs.get("file_cache")
        cache: dict[Path, str] | None = fc if isinstance(fc, dict) else None
        file_to_module, module_to_files, file_to_imports = self._build_import_index(all_candidate_files, repo_root, cache)

        changed_set = set(changed_files)
        discovered: set[Path] = set()
        frontier = set(py_changed)

        for _depth in range(self._DISCOVERY_MAX_DEPTH):
            next_frontier = self._expand_frontier(
                frontier,
                changed_set,
                discovered,
                file_to_imports,
                file_to_module,
                module_to_files,
                repo_root,
            )
            frontier = next_frontier
            if not frontier:
                break

        return sorted(discovered)

    @staticmethod
    def _expand_frontier(
        frontier: set[Path],
        changed_set: set[Path],
        discovered: set[Path],
        file_to_imports: dict[Path, set[str]],
        file_to_module: dict[Path, str],
        module_to_files: dict[str, list[Path]],
        repo_root: Path | None,
    ) -> set[Path]:
        next_frontier: set[Path] = set()
        for f in frontier:
            f_imports = file_to_imports.get(f)
            if f_imports is None:
                try:
                    content = f.read_text(encoding="utf-8")
                    f_imports = _extract_imports_from_content(content, f, repo_root)
                except (OSError, UnicodeDecodeError):
                    continue

            for imp in f_imports:
                for target in module_to_files.get(imp, []):
                    if target not in changed_set and target not in discovered:
                        discovered.add(target)
                        next_frontier.add(target)

            f_module = file_to_module.get(f) or path_to_module(f, repo_root)
            if f_module:
                for candidate, cand_imports in file_to_imports.items():
                    if candidate not in changed_set and candidate not in discovered and f_module in cand_imports:
                        discovered.add(candidate)
                        next_frontier.add(candidate)

        return next_frontier

    def _build_import_index(
        self,
        all_candidate_files: list[Path],
        repo_root: Path | None,
        file_cache: dict[Path, str] | None = None,
    ) -> tuple[dict[Path, str], dict[str, list[Path]], dict[Path, set[str]]]:
        file_to_module: dict[Path, str] = {}
        module_to_files: dict[str, list[Path]] = defaultdict(list)

        for f in all_candidate_files:
            if not _is_python_file(f):
                continue
            module = path_to_module(f, repo_root)
            if module:
                file_to_module[f] = module
                module_to_files[module].append(f)
                parts = module.split(".")
                for i in range(1, len(parts)):
                    prefix = ".".join(parts[:i])
                    module_to_files[prefix].append(f)

        self._enrich_with_reexports(
            all_candidate_files,
            repo_root,
            file_cache,
            file_to_module,
            module_to_files,
        )

        if _RG_BIN and repo_root:
            candidate_set = {f for f in all_candidate_files if _is_python_file(f)}
            try:
                file_to_imports = _build_import_index_rg(repo_root, candidate_set)
                return file_to_module, dict(module_to_files), dict(file_to_imports)
            except (subprocess.TimeoutExpired, OSError):
                logger.debug("ripgrep import index failed, falling back to ast")

        file_to_imports_ast: dict[Path, set[str]] = {}
        for f in all_candidate_files:
            if not _is_python_file(f):
                continue
            content = file_cache.get(f) if file_cache else None
            if content is None:
                try:
                    content = f.read_text(encoding="utf-8")
                except (OSError, UnicodeDecodeError):
                    continue
            file_to_imports_ast[f] = _extract_imports_from_content(content, f, repo_root)

        return file_to_module, dict(module_to_files), file_to_imports_ast

    @staticmethod
    def _enrich_with_reexports(
        all_candidate_files: list[Path],
        repo_root: Path | None,
        file_cache: dict[Path, str] | None,
        file_to_module: dict[Path, str],
        module_to_files: dict[str, list[Path]],
    ) -> None:
        init_files = [f for f in all_candidate_files if f.name == _INIT_PY]
        for init_f in init_files:
            pkg_module = path_to_module(init_f, repo_root)
            if not pkg_module:
                continue
            reexports = _resolve_init_reexports(init_f, repo_root, file_cache)
            existing = set(module_to_files.get(pkg_module, []))
            for _name, source_path in reexports.items():
                if source_path not in existing:
                    module_to_files[pkg_module].append(source_path)
                    existing.add(source_path)
                    if source_path not in file_to_module:
                        src_module = path_to_module(source_path, repo_root)
                        if src_module:
                            file_to_module[source_path] = src_module

    @staticmethod
    def _enrich_frags_with_reexports(
        py_frags: list[Fragment],
        repo_root: Path | None,
        file_cache: dict[Path, str],
        module_to_frags: dict[str, list[FragmentId]],
        path_to_frags: dict[Path, list[FragmentId]],
    ) -> None:
        for f in py_frags:
            if f.path.name != _INIT_PY:
                continue
            pkg_module = path_to_module(f.path, repo_root)
            if not pkg_module:
                continue
            reexports = _resolve_init_reexports(f.path, repo_root, file_cache)
            existing = set(module_to_frags.get(pkg_module, []))
            for _name, source_path in reexports.items():
                for sf in path_to_frags.get(source_path, []):
                    if sf not in existing:
                        module_to_frags[pkg_module].append(sf)

    @staticmethod
    def _build_frag_indexes(
        py_frags: list[Fragment],
        info_cache: dict[FragmentId, PyFragmentInfo],
        repo_root: Path | None,
    ) -> tuple[
        dict[str, list[FragmentId]], dict[FragmentId, frozenset[str]], dict[str, list[FragmentId]], dict[Path, list[FragmentId]]
    ]:
        name_to_defs: dict[str, list[FragmentId]] = defaultdict(list)
        frag_defines: dict[FragmentId, frozenset[str]] = {}
        for f in py_frags:
            info = info_cache[f.id]
            frag_defines[f.id] = info.defines
            for name in info.defines:
                name_to_defs[name].append(f.id)

        module_to_frags: dict[str, list[FragmentId]] = defaultdict(list)
        path_to_frags: dict[Path, list[FragmentId]] = defaultdict(list)
        for f in py_frags:
            module = path_to_module(f.path, repo_root)
            if module:
                module_to_frags[module].append(f.id)
            path_to_frags[f.path].append(f.id)

        return name_to_defs, frag_defines, module_to_frags, path_to_frags

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        py_frags = [f for f in fragments if _is_python_file(f.path)]
        if not py_frags:
            return {}

        info_cache: dict[FragmentId, PyFragmentInfo] = {}
        for f in py_frags:
            info_cache[f.id] = analyze_python_fragment(f.content)

        name_to_defs, frag_defines, module_to_frags, path_to_frags = self._build_frag_indexes(
            py_frags,
            info_cache,
            repo_root,
        )

        file_cache: dict[Path, str] = {f.path: f.content for f in py_frags}
        self._enrich_frags_with_reexports(py_frags, repo_root, file_cache, module_to_frags, path_to_frags)

        frag_imports = {f.id: _extract_imports_from_content(f.content, f.path, repo_root) for f in py_frags}
        frag_to_module = {f.id: m for f in py_frags if (m := path_to_module(f.path, repo_root))}

        edges: EdgeDict = {}
        for f in py_frags:
            info = info_cache[f.id]
            self_defs = set(frag_defines.get(f.id, frozenset()))
            src_imports = frag_imports.get(f.id, set())
            self._add_import_confirmed_edges(edges, f.id, info, name_to_defs, self_defs, src_imports, frag_to_module)
            self._add_import_edges(f, frag_imports[f.id], module_to_frags, edges)

        return edges

    _IMPORT_CONFIRMED_BOOST = 1.5
    _IMPORT_UNCONFIRMED_PENALTY = 0.2
    _IMPORT_WEIGHT = 0.75

    def _add_import_confirmed_edges(
        self,
        edges: EdgeDict,
        src_id: FragmentId,
        info: PyFragmentInfo,
        name_to_defs: dict[str, list[FragmentId]],
        self_defs: set[str],
        src_imports: set[str],
        frag_to_module: dict[FragmentId, str],
    ) -> None:
        for ref_set, base_weight in [
            (info.calls, _CALL_WEIGHT),
            (info.references, _SYMBOL_REF_WEIGHT),
            (info.type_refs, _TYPE_REF_WEIGHT),
        ]:
            for name in ref_set:
                if name in self_defs:
                    continue
                for dst_id in name_to_defs.get(name, []):
                    if dst_id == src_id:
                        continue
                    dst_module = frag_to_module.get(dst_id, "")
                    confirmed = bool(dst_module and dst_module in src_imports)
                    factor = self._IMPORT_CONFIRMED_BOOST if confirmed else self._IMPORT_UNCONFIRMED_PENALTY
                    w = base_weight * factor
                    edges[(src_id, dst_id)] = max(edges.get((src_id, dst_id), 0.0), w)
                    rev_w = w * self.reverse_weight_factor
                    edges[(dst_id, src_id)] = max(edges.get((dst_id, src_id), 0.0), rev_w)

    def _add_import_edges(
        self,
        frag: Fragment,
        imports: set[str],
        module_to_frags: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for imp in imports:
            targets = module_to_frags.get(imp, [])
            for tgt in targets:
                if tgt == frag.id:
                    continue
                edges[(frag.id, tgt)] = max(edges.get((frag.id, tgt), 0.0), self._IMPORT_WEIGHT)

File: diffctx/src/edges/semantic/python.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::extensions::PYTHON_EXTENSIONS;
use crate::config::weights::LANG_WEIGHTS;
use crate::types::{Fragment, FragmentId};

use super::super::base::{self, EdgeBuilder, path_to_module};
use super::super::EdgeDict;

fn is_python_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    PYTHON_EXTENSIONS.contains(ext.as_str())
}

static IMPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*import\s+([\w.]+)").unwrap());
static FROM_IMPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*from\s+([\w.]+)\s+import\s+(.+)").unwrap());
static CALL_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\b([A-Za-z_]\w*)\s*\(").unwrap());
static TYPE_REF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\b([A-Z]\w*)\b").unwrap());
static DEF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*(?:def|class|async\s+def)\s+([A-Za-z_]\w*)").unwrap());

fn extract_imports(content: &str, path: &Path, repo_root: Option<&Path>) -> FxHashSet<String> {
    let mut imports = FxHashSet::default();
    for cap in IMPORT_RE.captures_iter(content) {
        imports.insert(cap[1].to_string());
    }
    for cap in FROM_IMPORT_RE.captures_iter(content) {
        let module = &cap[1];
        if module.starts_with('.') {
            if let Some(parent) = path.parent() {
                let module_path = path_to_module(parent, repo_root);
                if !module_path.is_empty() {
                    imports.insert(module_path);
                }
            }
        } else {
            imports.insert(module.to_string());
            let parts: Vec<&str> = module.split('.').collect();
            for i in 1..parts.len() {
                imports.insert(parts[..i].join("."));
            }
        }
    }
    imports
}

fn extract_defines(content: &str) -> FxHashSet<String> {
    DEF_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect()
}

fn extract_calls(content: &str) -> FxHashSet<String> {
    CALL_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .filter(|n| !PY_KEYWORDS.contains(n.as_str()))
        .collect()
}

fn extract_type_refs(content: &str) -> FxHashSet<String> {
    TYPE_REF_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect()
}

static PY_KEYWORDS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [
        "if", "for", "while", "return", "def", "class", "import", "from", "as", "with", "try",
        "except", "finally", "raise", "pass", "break", "continue", "yield", "lambda", "assert",
        "del", "elif", "else", "global", "nonlocal", "and", "or", "not", "is", "in", "async",
        "await", "True", "False", "None", "print", "len", "range", "type", "list", "dict", "set",
        "tuple", "str", "int", "float", "bool", "super", "isinstance", "hasattr", "getattr",
        "setattr", "property", "staticmethod", "classmethod",
    ]
    .iter()
    .copied()
    .collect()
});

const IMPORT_WEIGHT: f64 = 0.75;
const IMPORT_CONFIRMED_BOOST: f64 = 1.5;
const IMPORT_UNCONFIRMED_PENALTY: f64 = 0.2;
const REVERSE_FACTOR: f64 = 0.5;

pub struct PythonEdgeBuilder;

impl EdgeBuilder for PythonEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let py_frags: Vec<&Fragment> = fragments.iter().filter(|f| is_python_file(Path::new(f.path()))).collect();
        if py_frags.is_empty() {
            return FxHashMap::default();
        }

        let weights = LANG_WEIGHTS.get("python").expect("python weights");
        let call_weight = weights.call;
        let symbol_ref_weight = weights.symbol_ref;
        let type_ref_weight = weights.type_ref;

        let mut name_to_defs: FxHashMap<String, Vec<FragmentId>> = FxHashMap::default();
        let mut frag_defines: FxHashMap<FragmentId, FxHashSet<String>> = FxHashMap::default();
        let mut module_to_frags: FxHashMap<String, Vec<FragmentId>> = FxHashMap::default();

        for f in &py_frags {
            let defines = extract_defines(&f.content);
            for name in &defines {
                name_to_defs.entry(name.clone()).or_default().push(f.id.clone());
            }
            frag_defines.insert(f.id.clone(), defines);

            let module = path_to_module(Path::new(f.path()), repo_root);
            if !module.is_empty() {
                module_to_frags.entry(module).or_default().push(f.id.clone());
            }
        }

        let frag_imports: FxHashMap<FragmentId, FxHashSet<String>> = py_frags
            .iter()
            .map(|f| {
                let imports = extract_imports(&f.content, Path::new(f.path()), repo_root);
                (f.id.clone(), imports)
            })
            .collect();

        let frag_to_module: FxHashMap<FragmentId, String> = py_frags
            .iter()
            .filter_map(|f| {
                let m = path_to_module(Path::new(f.path()), repo_root);
                if m.is_empty() { None } else { Some((f.id.clone(), m)) }
            })
            .collect();

        let mut edges: EdgeDict = FxHashMap::default();

        for f in &py_frags {
            let self_defs = frag_defines.get(&f.id).cloned().unwrap_or_default();
            let src_imports = frag_imports.get(&f.id).cloned().unwrap_or_default();

            let calls = extract_calls(&f.content);
            let type_refs = extract_type_refs(&f.content);
            let refs: FxHashSet<String> = f.identifiers.iter()
                .filter(|id| !self_defs.contains(*id))
                .cloned()
                .collect();

            for (ref_set, base_weight) in [
                (&calls, call_weight),
                (&refs, symbol_ref_weight),
                (&type_refs, type_ref_weight),
            ] {
                for name in ref_set {
                    if self_defs.contains(name) {
                        continue;
                    }
                    if let Some(dst_ids) = name_to_defs.get(name) {
                        for dst_id in dst_ids {
                            if dst_id == &f.id {
                                continue;
                            }
                            let dst_module = frag_to_module.get(dst_id).map(|s| s.as_str()).unwrap_or("");
                            let confirmed = !dst_module.is_empty() && src_imports.contains(dst_module);
                            let factor = if confirmed { IMPORT_CONFIRMED_BOOST } else { IMPORT_UNCONFIRMED_PENALTY };
                            let w = base_weight * factor;
                            let key_fwd = (f.id.clone(), dst_id.clone());
                            let existing = edges.get(&key_fwd).copied().unwrap_or(0.0);
                            if w > existing {
                                edges.insert(key_fwd, w);
                            }
                            let rev_w = w * REVERSE_FACTOR;
                            let key_rev = (dst_id.clone(), f.id.clone());
                            let existing_rev = edges.get(&key_rev).copied().unwrap_or(0.0);
                            if rev_w > existing_rev {
                                edges.insert(key_rev, rev_w);
                            }
                        }
                    }
                }
            }

            for imp in &src_imports {
                if let Some(targets) = module_to_frags.get(imp) {
                    for tgt in targets {
                        if tgt == &f.id {
                            continue;
                        }
                        let key = (f.id.clone(), tgt.clone());
                        let existing = edges.get(&key).copied().unwrap_or(0.0);
                        if IMPORT_WEIGHT > existing {
                            edges.insert(key, IMPORT_WEIGHT);
                        }
                    }
                }
            }
        }

        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let py_changed: Vec<&PathBuf> = changed.iter().filter(|f| is_python_file(f)).collect();
        if py_changed.is_empty() {
            return vec![];
        }

        let mut file_to_module: FxHashMap<PathBuf, String> = FxHashMap::default();
        let mut module_to_files: FxHashMap<String, Vec<PathBuf>> = FxHashMap::default();
        let mut file_to_imports: FxHashMap<PathBuf, FxHashSet<String>> = FxHashMap::default();

        for f in candidates {
            if !is_python_file(f) {
                continue;
            }
            let module = path_to_module(f, repo_root);
            if !module.is_empty() {
                file_to_module.insert(f.clone(), module.clone());
                module_to_files.entry(module.clone()).or_default().push(f.clone());
                let parts: Vec<&str> = module.split('.').collect();
                for i in 1..parts.len() {
                    module_to_files
                        .entry(parts[..i].join("."))
                        .or_default()
                        .push(f.clone());
                }
            }
            let content = base::read_file_cached(f, file_cache);
            if let Some(c) = content {
                file_to_imports.insert(f.clone(), extract_imports(&c, f, repo_root));
            }
        }

        let changed_set: FxHashSet<PathBuf> = changed.iter().cloned().collect();
        let mut discovered: FxHashSet<PathBuf> = FxHashSet::default();
        let mut frontier: FxHashSet<PathBuf> = py_changed.iter().map(|f| (*f).clone()).collect();

        for _ in 0..2 {
            let mut next_frontier: FxHashSet<PathBuf> = FxHashSet::default();
            for f in &frontier {
                let f_imports = file_to_imports.get(f).cloned().unwrap_or_default();
                for imp in &f_imports {
                    if let Some(targets) = module_to_files.get(imp) {
                        for target in targets {
                            if !changed_set.contains(target) && !discovered.contains(target) {
                                discovered.insert(target.clone());
                                next_frontier.insert(target.clone());
                            }
                        }
                    }
                }
                let f_module = file_to_module
                    .get(f)
                    .cloned()
                    .unwrap_or_else(|| path_to_module(f, repo_root));
                if !f_module.is_empty() {
                    for (candidate, cand_imports) in &file_to_imports {
                        if !changed_set.contains(candidate)
                            && !discovered.contains(candidate)
                            && cand_imports.contains(&f_module)
                        {
                            discovered.insert(candidate.clone());
                            next_frontier.insert(candidate.clone());
                        }
                    }
                }
            }
            if next_frontier.is_empty() {
                break;
            }
            frontier = next_frontier;
        }

        let mut result: Vec<PathBuf> = discovered.into_iter().collect();
        result.sort();
        result
    }
}
